### ISSUE TITLE
Env variable interpolation stops working when extending

### ISSUE THREAD

**Issue by indietyp** [sagiegurari/cargo-make#685]
### Describe The Bug

When redefining environmental variables, from a Makefile that we extend from and execute any task `@@split()` no longer works.

### To Reproduce

Minimal Reproducible Example:

extend.toml

```toml
extend = { path = "base.toml" }

[env]
CARGO_TEST_HACK_FLAGS_BASE = "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow"
CARGO_TEST_HACK_FLAGS = "${CARGO_TEST_HACK_FLAGS_BASE} --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small"


[tasks.test2]
command = "echo"
args = ["@@split(CARGO_TEST_HACK_FLAGS, )"]
```

base.toml

```toml
[config]
default_to_workspace = false

[config.modify_core_tasks]
private = true
namespace = "default"

[env]
CARGO_TEST_HACK_FLAGS = "--workspace --feature-powerset"


[tasks.test1]
command = "echo"
args = ["@@split(CARGO_TEST_HACK_FLAGS, )"]
```

```console
> cargo make --makefile extend.toml test2
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: extend.toml
[cargo-make] INFO - Task: test2
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: default::legacy-migration
[cargo-make] INFO - Execute Command: "echo" "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small"
--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small
[cargo-make] INFO - Build Done in 0.10 seconds.
> cargo make --makefile extend.toml test1
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: extend.toml
[cargo-make] INFO - Task: test1
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: default::legacy-migration
[cargo-make] INFO - Execute Command: "echo" "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small"
--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small
[cargo-make] INFO - Build Done in 0.11 seconds.
```

**Comment by indietyp** [comment]
Update: if `CARGO_TEST_HACK_FLAGS_BASE` is defined in `base.toml` This issue does not occur:

extend.toml: unchanged

base.toml

```toml
[config]
default_to_workspace = false

[config.modify_core_tasks]
private = true
namespace = "default"

[env]
CARGO_TEST_HACK_FLAGS_BASE = ""
CARGO_TEST_HACK_FLAGS = "--workspace --feature-powerset"


[tasks.test1]
command = "echo"
args = ["@@split(CARGO_TEST_HACK_FLAGS, )"]
```

```console
> cargo make --makefile extend.toml test2
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: extend.toml
[cargo-make] INFO - Task: test2
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: default::legacy-migration
[cargo-make] INFO - Execute Command: "echo" "--workspace" "--feature-powerset" "--exclude-features" "default" "--optional-deps" "eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small"
--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow --group-features anyhow,eyre --group-features std,hooks --group-features spantrace,futures,small
[cargo-make] INFO - Build Done in 0.11 seconds.
```

**Comment by sagiegurari** [comment]
The order to env vars is preserved since env vars may depend on others (as you did for example).
So order is defined from base to extended and so on....
in your case its as follows:

base->CARGO_TEST_HACK_FLAGS
extended->CARGO_TEST_HACK_FLAGS_BASE, CARGO_TEST_HACK_FLAGS

so overall order is:
CARGO_TEST_HACK_FLAGS, CARGO_TEST_HACK_FLAGS_BASE

so when CARGO_TEST_HACK_FLAGS is evaluated, CARGO_TEST_HACK_FLAGS_BASE is not yet defined and can't be used and replaced with empty string.

I know it feels strange and unexpected, but order is important and must persist.

**Comment by indietyp** [comment]
Thanks for the reply. This is pretty unintuitive because of the following:
The value isn't replaced with an empty string but instead with the value that isn't split.


What I expect:

```
"echo" "--workspace" "--feature-powerset" "--exclude-features" "default" "--optional-deps" "eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small"
```

What I get:

```
"echo" "--workspace --feature-powerset --exclude-features default --optional-deps eyre,anyhow" "--group-features" "anyhow,eyre" "--group-features" "std,hooks" "--group-features" "spantrace,futures,small"
```

As you can see, the value is still inserted, but only as a string, which then isn't split. I'd say this is quite unintuitive.



In my mind, how it would work, ideally, is doing it kind of like this:

```rust
let CARGO_TEST_HACK_FLAGS = "x";
let CARGO_TEST_HACK_FLAGS_BASE = "x";
let CARGO_TEST_HACK_FLAGS = "{CARGO_TEST_HACK_FLAGS_BASE} x"
```

So that instead of finding only the first occurrence, being able to redefine env variables in extended Makefiles, which then have their dependency graph. Otherwise, it would not be obvious.

I hope that makes sense.

**Comment by sagiegurari** [comment]
@indietyp i get the confusion. it is confusing. but i'll give you a counter scenario

base
env1 = x
env2 = ${env1}

extended
env1 = y

with your solution it will do
env1 = x
env2 = x
env1 = y

with mine
env1 = y
env2 = y

so order matters a lot and extend doesn't change execution order and doesn't cause multiple executions.
its taking multiple configs and putting them one on top of the other and running it once the end result.

which one makes more sense? it depends on the scenario. but i felt the current way is more logical since we are talking about configs and not code.

**Comment by indietyp** [comment]
That's also what I expect, but my case is

base
env2 = y

extended
env1 = y
env2 = ${env1}

That env2 is also y, which in the current case works, but not really.

**Comment by sagiegurari** [comment]
i gave different scenario of possible issue in case i go with your suggestion. please read carefully.

**Comment by indietyp** [comment]
Sorry, I misunderstood. I see what you mean now. This makes sense. It's all about trade-offs. I am still confused about why the string is inserted into the arguments. Do you have any idea? Maybe a diagnostic warning about this specific case would be suitable?

**Comment by sagiegurari** [comment]
i can't remember if i'm changing it to empty string or leaving it as is when i can't find the env var. is that the case you are talking about?
can you give basic example for me to check?

**Comment by indietyp** [comment]
I think I finally know where the confusion is coming from. With my suggestion

base
env1 = x
env2 = ${env1}

extended
env1 = y

result
env1 = y
env2 = y

would still work, but also


base
env1 = x
env2 = x

extended
env1 = y
env2 = ${env1}

result
env1 = y
env2 = y

would work because when we encounter the 2nd `env2`, we recalculate the "dependency" graph. In the first example, `env1` is already defined when we use `env2`, during reconciliation we then take the current value from `env1` (overwritten in extended) instead of `env2`. In the 2nd example, we **redefine** `env2`, so treat it as a completely new env variable, where we then recalculate the "dependency" graph, taking into account all previously defined env variables. Instead of taking the dependency graph, we previously used for `env2`. Does that make sense?

This would make the behavior more consistent and not change any existing behavior. That's what I have been trying to explain.

This could be explained through the following pseudo algorithm:

```
create a graph G where every node is an environmental variable
for every environmental variable in base and extended {
  clear all previous edges to "environmental" variable
  for every mentioned variable {
    create an edge "environmental variable" -> "mentioned variable"
  }
}

check for cycles in graph `G`. If there are any cycles/strongly connected components, abort.
Create a topological sorting of all nodes
reverse the topological sorting
for every variable {
  search for the latest value of variable
  evaluate and store as result
}
```

I prepared a gist that elaborates a bit further: https://gist.github.com/indietyp/5f78da9129a7dcaf140445b20d4e83c7

This is ofc not ideal. For the sake of brevity, this code omits some things like optimizations (I think one can safely just use the latest value in the graph analysis phase).

This does two things actually: it allows for both use cases (will evaluate both to `y`, `y`) and no longer needs a specific order, so this is order independent. (so one can mention `env2` in the first example before `env1` without breaking anything).

**Comment by indietyp** [comment]
> i can't remember if i'm changing it to empty string or leaving it as is when i can't find the env var. is that the case you are talking about?
> can you give basic example for me to check?

In the example I already gave in https://github.com/sagiegurari/cargo-make/issues/685#issuecomment-1206660949, the variable is inserted (look at the echo arguments), but isn't split.

**Comment by sagiegurari** [comment]
> I think I finally know where the confusion is coming from. With my suggestion
> 
> base env1 = x env2 = ${env1}
> 
> extended env1 = y
> 
> result env1 = y env2 = y
> 
> would still work, but also
> 
> base env1 = x env2 = x
> 
> extended env1 = y env2 = ${env1}
> 
> result env1 = y env2 = y

i'm pretty sure this is how it works right now.
i think you meant to suggest:

```
base
# NOT DEFINE -> env1 = x
env2 = x

extended
env1 = y
env2 = ${env1}

result
env1 = y
env2 = y
```

right?

anyhow, while your suggestion (if i understood correctly) is a good one, you must understand the current implementation

1. makefile loading -> just loads the makefile while keeping env vars in order. it doesn't know/understand the env vars at all.
2. setup env -> this one is pretty complex with multiple layers and it does the job. BUT its too late per your suggestion as we no longer know how to 'fix' dependencies. so we need to merge this into step 1.
3. setup env per task -> the whole process is also happening per task for env vars defined on the task level!!! so if we want to be consistent, we need to do this here, so step 1 would also need to know how tasks relate to each other (think extend, clear, namespaces and platform overrides, ... mess)

so.... it feels like an overkill when you can just add to the base toml ```myvar=``` and thats it.
yes! it is confusing! thats why we are talking so much about it :D and I do want to make sure we provide best solution. but i need to maintain that code later on and understand what the hack is going on and that i can move fast and give more capabilities.
so i'm not sure i'm going to go for it.

**Comment by indietyp** [comment]
> right?

Yes! that is what I want, but this isn't working as intended (at least from what I understood)

```
base
# NOT DEFINE -> env1 = x
env2 = x y z

extended
env1 = a b c
env2 = ${env1} x y z

result
env1 = a b c
env2 = a b c x y z

@@split(env2) = "a" "b" "c" "x" "y" "z"
# CURRENTLY @@split(env2) = "a b c" "x" "y" "z"
# ^ this is the confusing part
```

> so.... it feels like an overkill when you can just add to the base toml `myvar=` and thats it.

This is certainly a fix for the problem, tho it causes some confusion. (which I will probably implement). Thanks for the elaboration on how `cargo-make` works. I think my suggestion would still add something, but I can imagine that integrating it correctly could be a real mess. (would boil down to deferring environmental evaluation until the end in `3`), collecting the environment variables along the way.

If I find the time to do so - would you be ok w/ me opening a PR, trying to implement my suggestion? I'd love to help out, and we'd be able to discuss implementation and such further in the PR.

**Comment by sagiegurari** [comment]
I'm very happy to get PRs.
This specific issue might be really messy so I wouldn't invest too much in it.
If you got simple solution, i'll be happy to accept it, but if it is going to be as complex as i suspect, i would probably not.
So, just making sure you are aware of it before you invest too much time.

in general, i'm accepting PRs into cargo-make (there are over 40 contributors)

**Comment by sagiegurari** [comment]
@indietyp thanks for the PR, this is now released as part of cargo-make 0.36.0
